runtime/src/snapshot_utils.rs

```diff
@@ -665,81 +1011,209 @@ pub fn bank_from_snapshot_archive<P>(
+            }),
         frozen_account_pubkeys,
-        &unpacked_snapshots_dir,
         account_paths,
         unpacked_append_vec_map,
         genesis_config,
         debug_keys,
         additional_builtins,
-        account_indexes,
+        account_secondary_indexes,
         accounts_db_caching_enabled,
         limit_load_slot_count_from_snapshot,
         shrink_ratio,
         verify_index,
     )?;
-    measure.stop();
+    measure_rebuild.stop();
+    info!("{}", measure_rebuild);
 
-    let mut verify = Measure::start("verify");
+    let mut measure_verify = Measure::start("verify");
     if !bank.verify_snapshot_bank(test_hash_calculation)
         && limit_load_slot_count_from_snapshot.is_none()
     {
         panic!("Snapshot bank for slot {} failed to verify", bank.slot());
     }
-    verify.stop();
+    measure_verify.stop();
+
     let timings = BankFromArchiveTimings {
-        rebuild_bank_from_snapshots_us: measure.as_us(),
-        untar_us: untar.as_us(),
-        verify_snapshot_bank_us: verify.as_us(),
+        rebuild_bank_from_snapshots_us: measure_rebuild.as_us(),
+        full_snapshot_untar_us: unarchived_full_snapshot.measure_untar.as_us(),
+        incremental_snapshot_untar_us: unarchived_incremental_snapshot
+            .map_or(0, |unarchive_preparation_result| {
+                unarchive_preparation_result.measure_untar.as_us()
+            }),
+        verify_snapshot_bank_us: measure_verify.as_us(),
     };
-
     Ok((bank, timings))
 }
 
-/// Build the snapshot archive path from its components: the snapshot archive output directory, the
+/// Perform the common tasks when unarchiving a snapshot.  Handles creating the temporary
+/// directories, untaring, reading the version file, and then returning those fields plus the
+/// unpacked append vec map.
+fn unarchive_snapshot<P, Q>(
+    snapshots_dir: P,
+    unpacked_snapshots_dir_prefix: &'static str,
+    snapshot_archive_path: Q,
+    measure_name: &'static str,
+    account_paths: &[PathBuf],
+    archive_format: ArchiveFormat,
+    parallel_divisions: usize,
+) -> Result<UnarchivedSnapshot>
+where
+    P: AsRef<Path>,
+    Q: AsRef<Path>,
+{
+    let unpack_dir = tempfile::Builder::new()
+        .prefix(unpacked_snapshots_dir_prefix)
+        .tempdir_in(snapshots_dir)?;
+    let unpacked_snapshots_dir = unpack_dir.path().join("snapshots");
+
+    let mut measure_untar = Measure::start(measure_name);
+    let unpacked_append_vec_map = untar_snapshot_in(
+        snapshot_archive_path,
+        unpack_dir.path(),
+        account_paths,
+        archive_format,
+        parallel_divisions,
+    )?;
+    measure_untar.stop();
+    info!("{}", measure_untar);
+
+    let unpacked_version_file = unpack_dir.path().join("version");
+    let mut snapshot_version = String::new();
+    File::open(unpacked_version_file).and_then(|mut f| f.read_to_string(&mut snapshot_version))?;
```

Comment:
here

Reply:
also, I just noticed we're better off checking file size of `snapshot_version`? for eventual untrusted snapshot thing?

Otherwise, it could be dos vector if crafted snapshot contains a 100G zeroed snapshot_version file because we read all to memory first....

unrelated to this pr off course. so could be done as a follow up pr...

Reply:
ref: #7167

Reply:
New issue created: #18841